src/main/java/com/google/devtools/build/lib/worker/WorkerProxy.java

```diff
@@ -0,0 +1,106 @@
+// Copyright 2018 The Bazel Authors. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.devtools.build.lib.worker;
+
+import com.google.devtools.build.lib.sandbox.SandboxHelpers;
+import com.google.devtools.build.lib.vfs.Path;
+import com.google.devtools.build.lib.vfs.PathFragment;
+import java.io.ByteArrayInputStream;
+import java.io.ByteArrayOutputStream;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.io.IOException;
+import java.util.Map;
+import java.util.Set;
+
+/**
+ * A proxy that talks to the multiplexers
+ */
+final class WorkerProxy extends Worker {
+  private WorkerMultiplexer workerMultiplexer;
+  private ByteArrayOutputStream request;
+  private Thread shutdownHook;
+
+  WorkerProxy(WorkerKey workerKey, int workerId, Path workDir, Path logFile) {
+    super(workerKey, workerId, workDir, logFile);
+    this.workerMultiplexer = WorkerMultiplexer.getInstance(workerKey.hashCode());
+    this.request = new ByteArrayOutputStream();
+
+    final Worker self = this;
+    this.shutdownHook =
+      new Thread(
+        () -> {
+          try {
+            self.shutdownHook = null;
+            self.destroy();
+          } catch (IOException e) {
+            // We can't do anything here.
+          }
+        });
+    Runtime.getRuntime().addShutdownHook(shutdownHook);
+  }
+
+  @Override
+  void createProcess() throws IOException {
+    workerMultiplexer.createProcess(workerKey, workDir, logFile);
+  }
+
+  @Override
+  boolean isAlive() {
+    // This is horrible, but Process.isAlive() is only available from Java 8 on and this is the
+    // best we can do prior to that.
+    return workerMultiplexer.isProcessAlive();
+  }
+
+  @Override
+  public void prepareExecution(
+      Map<PathFragment, Path> inputFiles, SandboxHelpers.SandboxOutputs outputs, Set<PathFragment> workerFiles)
+      throws IOException {
+    createProcess();
+  }
+
+  @Override
+  synchronized void destroy() {
+    if (shutdownHook != null) {
```

Comment:
This can never be false. Something smells fishy.

Reply:
Definitely something wrong with this pattern - I'm consistently seeing:
```
2019-04-28T19:38:51.912870956Z ERROR: bazel crash in async thread:
2019-04-28T19:38:51.912890034Z java.lang.IllegalStateException: Shutdown in progress
2019-04-28T19:38:51.91289635Z 	at java.base/java.lang.ApplicationShutdownHooks.remove(ApplicationShutdownHooks.java:82)
2019-04-28T19:38:51.912916811Z 	at java.base/java.lang.Runtime.removeShutdownHook(Runtime.java:243)
2019-04-28T19:38:51.912924616Z 	at com.google.devtools.build.lib.worker.WorkerProxy.destroy(WorkerProxy.java:81)
2019-04-28T19:38:51.912929454Z 	at com.google.devtools.build.lib.worker.Worker.lambda$new$0(Worker.java:76)
2019-04-28T19:38:51.912934203Z 	at java.base/java.lang.Thread.run(Thread.java:834)
```

Reply:
Ah, I see the issue. The base class, `Worker`'s, shutdown hook is invoked which sets the **private** variable on `Worker` to `null` and then calls `self.destroy`, which invokes this method since it overrides the base class method. Then line 81 is executed because `WorkerProxy`'s `shutdownHook` is not null. Sorry if that's not clear, maybe explaining the fix will help:

1/ Remove :48-:58 in this class. It's already handled in the base class.
2/ Remove :80-:82 in this class. Add a super call to `destroy` and let the base class handle it.

Reply:
Good catch, I was also confused about this part. Fixed.